src/dvc_data/build.py

```diff
@@ -76,6 +78,74 @@ def _build_file(path, fs, name, odb=None, upload_odb=None, dry_run=False):
     return meta, obj
 
 
+def _find_files(path, fs, ignore=None):
+    if ignore:
+        walk_iter = ignore.walk(fs, path)
+    else:
+        walk_iter = fs.walk(path)
+
+    for root, _, fnames in walk_iter:
+        if DefaultIgnoreFile in fnames:
+            raise IgnoreInCollectedDirError(
+                DefaultIgnoreFile, fs.path.join(root, DefaultIgnoreFile)
+            )
+
+        # NOTE: we know for sure that root starts with path, so we can use
+        # faster string manipulation instead of a more robust relparts()
+        rel_key: Tuple[Optional[Any], ...] = ()
+        if root != path:
+            rel_key = tuple(root[len(path) + 1 :].split(fs.sep))
+
+        # NOTE: empty `fname` might happen with s3/gs/azure/etc, where empty
+        # objects like `dir/` might be used to create an empty dir
+        yield from (
+            (root + fs.sep + fname, (*rel_key, fname))
+            for fname in fnames
+            if fname
+        )
+
+
+def _diff_state_to_dir(path, fs, ignore=None, state=None, pbar=None):
+    files = dict(_find_files(path, fs, ignore=ignore))
+    total = len(files)
+
+    if pbar is not None:
+        pbar.total = total
+        pbar.refresh()
+
+    if not state:
+        return files, {}
+
+    new = {}
+    unchanged = {}
+    for file, (meta, hash_info) in state.get_many(files.keys(), fs):
+        key = files[file]
+        if not hash_info:
+            new[file] = key
+            continue
+
+        if hash_info:
+            unchanged[key] = (meta, hash_info)
+        if pbar:
+            pbar.update()
+    return new, unchanged
+
+
+def _build_new_files(paths, fs, name, odb, pbar=None, **kwargs):
+    new_entries = {}
+    hashes = {}
+    for path, key in paths.items():
+        meta, obj = _build_file(path, fs, name, odb, state=None, **kwargs)
+        new_entries[key] = (meta, obj.hash_info)
+        hashes[path] = obj.hash_info
+        if pbar is not None:
+            pbar.update()
+
+    if hashes and (state := odb.state):
+        state.save_many(hashes.items(), fs)
```

Comment:
what's the point of this complication? Let's keep things simple, this doesn't seem to be adding anything, unless I'm missing something.
```suggestion
    if hashes and odb.state:
        odb.state.save_many(hashes.items(), fs)
```